authentication/php/phpleague/src/PhpLeagueAccessTokenProvider.php

```diff
@@ -0,0 +1,115 @@
+<?php
+/**
+ * Copyright (c) Microsoft Corporation.  All Rights Reserved.
+ * Licensed under the MIT License.  See License in the project root
+ * for license information.
+ */
+
+
+namespace Microsoft\Kiota\Authentication;
+
+
+use Http\Promise\FulfilledPromise;
+use Http\Promise\Promise;
+use Http\Promise\RejectedPromise;
+use League\OAuth2\Client\Provider\GenericProvider;
+use League\OAuth2\Client\Token\AccessToken;
+use Microsoft\Kiota\Abstractions\Authentication\AccessTokenProvider;
+use Microsoft\Kiota\Abstractions\Authentication\AllowedHostsValidator;
+
+/**
+ * Class PhpLeagueAccessTokenProvider
+ * @package Microsoft\Kiota\Authentication
+ * @copyright 2022 Microsoft Corporation
+ * @license https://opensource.org/licenses/MIT MIT License
+ * @link https://developer.microsoft.com/graph
+ */
+class PhpLeagueAccessTokenProvider implements AccessTokenProvider
+{
+    /**
+     * @var TokenRequestContext {@link TokenRequestContext}
+     */
+    private TokenRequestContext $tokenRequestContext;
+    /**
+     * @var AllowedHostsValidator Validates whether a token should be fetched for a request url
+     */
+    private AllowedHostsValidator $allowedHostsValidator;
+    /**
+     * @var array<string, string>
+     */
+    private array $scopes;
+    /**
+     * @var AccessToken|null Token object to re-use before expiry
+     */
+    private ?AccessToken $cachedToken = null;
+    /**
+     * @var GenericProvider OAuth 2.0 provider from PHP League library
+     */
+    private GenericProvider $oauthProvider;
+
+    /**
+     * Creates a new instance
+     * @param TokenRequestContext $tokenRequestContext
+     * @param array $scopes
+     * @param array $allowedHosts
+     */
+    public function __construct(TokenRequestContext $tokenRequestContext, array $scopes, array $allowedHosts = [])
+    {
+
+        $this->tokenRequestContext = $tokenRequestContext;
+        if (empty($scopes)) {
+            throw new \InvalidArgumentException("Scopes cannot be empty");
+        }
+        $this->scopes = $scopes;
+        $this->allowedHostsValidator = new AllowedHostsValidator($allowedHosts);
+        $this->oauthProvider = new GenericProvider([
+            'urlAccessToken' => "https://login.microsoftonline.com/{$tokenRequestContext->getTenantId()}/oauth2/v2.0/token",
```

Comment:
I don't fully understand your comment. Please clarify how the metadata endpoint would be used for OBO?

(Side note)
I can use the metadata endpoints to fetch the token endpoint and dynamically init the GenericProvider. Would work well for the different National Clouds

Reply:
some libraries support both OAuth2 and OIDC. Supporting OIDC on top of OAuth means a couple of things:
- The lib can discover the token and auth code endpoints automatically (and the sign-out endpoint is inferred by convention I believe)
- The lib will perform token validation through the signature keys present in the metadata before using the token, or when used to secure the app itself (the later might only apply on OBO scenarios).
- The lib might be able to perform background token introspection to see whether the token has been invalidated by a third party and automatically trigger a cache invalidation, a token renewal instead of serving a now invalid token.
- The lib might be able to access additional user data through user info and expose that to the application context so it can be used by other components.
- There might be other benefits I'm not aware of

I don't know the library you're using, and I didn't take time to research its capability, but I'm suggesting that if it supports even only one of those benefits through OIDC, we might as well provide the endpoint so it can't do those things as well.

Reply:
Thanks for the clarification. No, the lib doesn't support any OIDC capabilities. I hadn't factored this into my considerations of libraries. Just seen an OIDC client library that i'll explore

Reply:
Similar update to previous comment